edb/pgsql/resolver/command.py

```diff
@@ -339,6 +369,409 @@ def _preprocess_insert_stmt(
+        ptr_ql.steps.append(qlast.Ptr(name='id'))
+
+        ptr_target = ptr.get_target(ctx.schema)
+        assert ptr_target
+        ptr_target_name: sn.Name = ptr_target.get_name(ctx.schema)
+        assert isinstance(ptr_target_name, sn.QualName)
+        ptr_ql = qlast.TypeCast(
+            type=qlast.TypeName(
+                maintype=qlast.ObjectRef(
+                    module=ptr_target_name.module,
+                    name=ptr_target_name.name,
+                )
+            ),
+            expr=ptr_ql,
+        )
+    return qlast.ShapeElement(
+        expr=qlast.Path(steps=[qlast.Ptr(name=ptr_name)]),
+        operation=qlast.ShapeOperation(op=qlast.ShapeOp.ASSIGN),
+        compexpr=ptr_ql,
+    )
+
+
+def _preprocess_insert_link_stmt(
+    stmt: pgast.InsertStmt,
+    link: s_links.Link,
+    sub_table: context.Table,
+    expected_columns: List[context.Column],
+    *,
+    ctx: Context,
+) -> PreprocessedDML:
+    if not any(c.name == 'source' for c in expected_columns):
+        raise errors.QueryError(
+            'column source is required when inserting into link tables',
+            span=stmt.span,
+        )
+    if not any(c.name == 'target' for c in expected_columns):
+        raise errors.QueryError(
+            'column target is required when inserting into link tables',
+            span=stmt.span,
+        )
+
+    sub_source = link.get_source(ctx.schema)
+    assert sub_source
+    sub_target = link.get_target(ctx.schema)
+    assert sub_target
+
+    # handle DEFAULT and prepare the value relation
+    value_relation, expected_columns = _preprocess_insert_value(
+        stmt.select_stmt, stmt.ctes, expected_columns
+    )
+
+    # if we are sure that we are inserting a single row
+    # we can skip for-loops and iterators, which produces better SQL
+    is_value_single = _has_at_most_one_row(stmt.select_stmt)
+
+    # prepare anchors for inserted value columns
+    value_name = ctx.alias_generator.get('ins_val')
+    iterator_name = ctx.alias_generator.get('ins_iter')
+    source_id = irast.PathId.from_type(
+        ctx.schema,
+        sub_source,
+        typename=sn.QualName('__derived__', value_name),
+        env=None,
+    )
+    link_ref = irtypeutils.ptrref_from_ptrcls(
+        schema=ctx.schema, ptrcls=link, cache=None, typeref_cache=None
+    )
+    value_id: irast.PathId = source_id.extend(ptrref=link_ref)
+
+    value_ql: qlast.PathElement = (
+        qlast.IRAnchor(name=value_name)
+        if is_value_single
+        else qlast.ObjectRef(name=iterator_name)
+    )
+
+    # a phantom relation that is supposed to hold the inserted value
+    # (in the resolver, this will be replaced by the real value relation)
+    value_cte_name = ctx.alias_generator.get('ins_value')
+    value_rel = pgast.Relation(name=value_cte_name)
+    value_columns: List[Tuple[str, bool]] = []
+    for expected_col in expected_columns:
+        if expected_col.name == 'source':
```

Comment:
Can these cases be unified more? I think there might be a cleanup possible, but no problem if not

Reply:
They *could* possibly be more unified, but I don't think they should.

There are many small differences. I've started with one large function with many if stmts to choose between the three cases, but it got way too messy, so I decided to have some simple bits of code duplicated for the sake of readability.